lib/netext/resolver.go

```diff
@@ -0,0 +1,135 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netext
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/lib/types"
+)
+
+// TODO: Figure out a non-global way to expose this for mocking in tests that
+// doesn't involve changing the Resolver interface...
+var LookupIP = net.LookupIP
+
+// Resolver is an interface that returns DNS information about a given host.
+type Resolver interface {
+	LookupIP(host string) (net.IP, error)
+}
+
+type resolver struct {
+	strategy   lib.DNSStrategy
+	rrm        *sync.Mutex
+	roundRobin map[string]uint8
+}
+
+type cacheRecord struct {
+	ips        []net.IP
+	lastLookup time.Time
+}
+
+type cacheResolver struct {
+	resolver
+	ttl   time.Duration
+	cm    *sync.Mutex
+	cache map[string]cacheRecord
+}
+
+// NewResolver returns a new DNS resolver. If ttl is a valid duration, responses
+// will be cached per host for the specified period. The IP returned from
+// LookupIP() will be selected based on the given strategy.
+func NewResolver(ttl types.NullDuration, strategy lib.DNSStrategy) Resolver {
+	var r Resolver
+	res := resolver{
+		strategy:   strategy,
+		rrm:        &sync.Mutex{},
+		roundRobin: make(map[string]uint8),
+	}
+	if ttl.Valid {
+		r = &cacheResolver{
+			resolver: res,
+			ttl:      time.Duration(ttl.Duration),
+			cm:       &sync.Mutex{},
+			cache:    make(map[string]cacheRecord),
+		}
+	} else {
+		r = &res
+	}
+
+	return r
```

Comment:
```suggestion
	res := resolver{
		strategy:   strategy,
		rrm:        &sync.Mutex{},
		roundRobin: make(map[string]uint8),
	}
	if !ttl.Valid || ttl.Duration == 0 {
		return &res
	}
	return &cacheResolver{
		resolver: res,
		ttl:      time.Duration(ttl.Duration),
		cm:       &sync.Mutex{},
		cache:    make(map[string]cacheRecord),
	}
```

Reply:
Hhmm yeah, that looks cleaner, and I could just pass `time.Duration` instead of our `types.NullDuration`.

Reply:
Done in cf9bbb2a, thanks.